lzero/mcts/buffer/game_buffer_ma.py

```diff
@@ -571,17 +716,265 @@ def _compute_target_policy_reanalyzed(self, policy_re_context: List[Any], model:
+            action_mask = [
+                list(np.ones(self._cfg.model.action_space_size, dtype=np.int8)) for _ in range(transition_batch_size)
+            ]
+            # NOTE: in continuous action space env: we set all legal_actions as -1
+            legal_actions = [
+                [-1 for _ in range(self._cfg.model.action_space_size)] for _ in range(transition_batch_size)
+            ]
+        else:
+            legal_actions = [[i for i, x in enumerate(action_mask[j]) if x == 1] for j in range(transition_batch_size)]
+
+        with torch.no_grad():
+            policy_obs_list = prepare_observation(policy_obs_list, self._cfg.model.model_type)
+            # split a full batch into slices of mini_infer_size: to save the GPU memory for more GPU actors
+            slices = int(np.ceil(transition_batch_size / self._cfg.mini_infer_size))
+            network_output = []
+            for i in range(slices):
+                beg_index = self._cfg.mini_infer_size * i
+                end_index = self._cfg.mini_infer_size * (i + 1)
+                m_obs = torch.from_numpy(policy_obs_list[beg_index:end_index]).to(self._cfg.device).float()
+                m_output = model.initial_inference(m_obs)
+
+                # 这段是干啥的？？？？？？？？？？？？？？？？？？？？？？？？？？？？？？
+                if not model.training:
+                    # if not in training, obtain the scalars of the value/reward
+                    [m_output.latent_state, m_output.value, m_output.policy_logits] = to_detach_cpu_numpy(
+                        [
+                            m_output.latent_state,
+                            inverse_scalar_transform(m_output.value, self._cfg.model.support_scale),
+                            m_output.policy_logits
+                        ]
+                    )
+
+                network_output.append(m_output)
+            
+
+
+            # 得到所有的obs对应的隐根节点，以及网络预测的policy和value
+            _, reward_pool, policy_logits_pool, latent_state_roots = concat_output(network_output, data_type='muzero')
+            reward_pool = reward_pool.squeeze().tolist()
+            policy_logits_pool = policy_logits_pool.tolist()
+            noises = [
+                np.random.dirichlet([self._cfg.root_dirichlet_alpha] * self._cfg.model.action_space_size
+                                    ).astype(np.float32).tolist() for _ in range(transition_batch_size)
+            ]
+
+            # 根据得到的这些latent 根节点搜索新的target_policy
+            # prepare之后构建新的roots类，并把现有的roots分批复制进去
+            # if self._cfg.mcts_ctree:
+            #     # cpp mcts_tree
+            #     roots = MCTSCtree.roots(transition_batch_size, legal_actions)
+            #     roots.prepare(self._cfg.root_noise_weight, noises, reward_pool, policy_logits_pool, to_play)
+            #     # do MCTS for a new policy with the recent target model
+            #     MCTSCtree(self._cfg).search(roots, model, latent_state_roots, to_play)
+            if self._cfg.mcts_ctree:
```

Comment:
def compute_all_filters(data, num_unroll_steps):
    data_by_iter = []
    for iter in range(num_unroll_steps + 1):
        iter_data = [x for i, x in enumerate(data)
                     if (i + 1) % (num_unroll_steps + 1) ==
                     ((num_unroll_steps + 1 - iter) % (num_unroll_steps + 1))]
        data_by_iter.append(iter_data)
    return data_by_iter

if self._cfg.mcts_ctree:
    legal_actions_by_iter = compute_all_filters(legal_actions, self._cfg.num_unroll_steps)
    noises_by_iter = compute_all_filters(noises, self._cfg.num_unroll_steps)
    reward_pool_by_iter = compute_all_filters(reward_pool, self._cfg.num_unroll_steps)
    policy_logits_pool_by_iter = compute_all_filters(policy_logits_pool, self._cfg.num_unroll_steps)
    to_play_by_iter = compute_all_filters(to_play, self._cfg.num_unroll_steps)
    latent_state_roots_by_iter = compute_all_filters(latent_state_roots, self._cfg.num_unroll_steps)
    true_action_by_iter = compute_all_filters(true_action, self._cfg.num_unroll_steps)

    temp_values = []
    temp_distributions = []
    mcts_ctree = MCTSCtree(self._cfg)

    for iter in range(self._cfg.num_unroll_steps + 1):
        iter_batch_size = transition_batch_size / (self._cfg.num_unroll_steps + 1)
        roots = MCTSCtree.roots(iter_batch_size, legal_actions_by_iter[iter])

        roots.prepare(self._cfg.root_noise_weight, 
                      noises_by_iter[iter], 
                      reward_pool_by_iter[iter],
                      policy_logits_pool_by_iter[iter], 
                      to_play_by_iter[iter])

        if iter == 0:
            mcts_ctree.search(roots, model, latent_state_roots_by_iter[iter], to_play_by_iter[iter])
        else:
            mcts_ctree.search_with_reuse(roots, model, latent_state_roots_by_iter[iter], 
                                         to_play_by_iter[iter],
                                         true_action_list=true_action_by_iter[iter], 
                                         reuse_value_list=iter_values)

        iter_values = roots.get_values()
        iter_distributions = roots.get_distributions()
        temp_values.append(iter_values)
        temp_distributions.append(iter_distributions)